pkg/node/node.go

```diff
@@ -424,41 +411,161 @@ func (n *Node) publishSignature(
 			continue
 		}
 
-		logger.Infof("signature submitted for keep [%s]", keepAddress.String())
+		if !(n.waitForSignature(keepAddress, digest) && n.confirmSignature(keepAddress, digest)) {
+			time.Sleep(retryDelay) // TODO: #413 Replace with backoff.
+			continue
+		}
+
 		return nil
 	}
 }
 
-// monitorKeepPublicKeySubmission observes the chain until either the first
-// conflicting public key is published or until keep established public key
-// or until key generation timed out.
-func (n *Node) monitorKeepPublicKeySubmission(
-	abort chan interface{},
+func (n *Node) waitForSignature(
 	keepAddress common.Address,
-) {
-	monitoringCtx, monitoringCancel := context.WithTimeout(
-		context.Background(),
-		monitorKeepPublicKeySubmissionTimeout,
+	digest [32]byte,
+) bool {
+	const waitTimeout = 10 * time.Minute
+	const checkTick = 1 * time.Minute
+
+	ctx, cancelCtx := context.WithTimeout(context.Background(), waitTimeout)
+	defer cancelCtx()
+
+	checkTicker := time.NewTicker(checkTick)
+	defer checkTicker.Stop()
+
+	logger.Infof(
+		"waiting for signature for keep [%s] to appear on-chain",
+		keepAddress.String(),
 	)
-	defer monitoringCancel()
 
-	publicKeyPublished := make(chan *eth.PublicKeyPublishedEvent)
-	conflictingPublicKey := make(chan *eth.ConflictingPublicKeySubmittedEvent)
+	for {
+		select {
+		case <-checkTicker.C:
+			// We check the public key periodically instead of relying on
+			// incoming events. The main motivation is that events could not be
+			// trusted here because they may come from a forked chain or
+			// the same event can be delivered multiple times.
+			isAwaitingSignature, err := n.ethereumChain.IsAwaitingSignature(
+				keepAddress,
+				digest,
+			)
+			if err != nil {
+				logger.Error(
+					"failed to perform signature check while waiting "+
+						"for signature for keep [%s]: [%v]",
+					keepAddress.String(),
+					err,
+				)
+				continue
+			}
 
-	subscriptionPublicKeyPublished, err := n.ethereumChain.OnPublicKeyPublished(
-		keepAddress,
-		func(event *eth.PublicKeyPublishedEvent) {
-			publicKeyPublished <- event
+			if !isAwaitingSignature {
+				logger.Infof(
+					"signature for keep [%s] appeared on-chain",
+					keepAddress.String(),
+				)
+				return true
+			}
+		case <-ctx.Done():
+			logger.Error(
+				"stop waiting for signature for keep [%s] because "+
+					"[%v] timeout has been exceeded",
+				keepAddress.String(),
+				waitTimeout,
+			)
```

Comment:
`Error->Errorf` - it's currently rendered as:
```
2020-11-20T14:42:25.031+0100	ERROR	keep-ecdsa	stop waiting for signature for keep [%s] because [%v] timeout has been exceeded0xe1C664a63BD1D017270aaE9bda45A776bA86a07810m0s
```

Let's also rework the message a little to be less confusing - we are stopping only the current iteration not the whole process. My proposition:
```
signature for keep [%s] has not appeared on the chain after [%v] from submitting it
```